File: ios/ShouldIGo/Features/Notifications/NotificationSheet.swift
import SwiftUI

struct NotificationSheet: View {
    let spot: Spot
    @ObservedObject var pushManager: PushManager
    @Binding var isPresented: Bool

    @State private var selectedThreshold = 65
    @State private var isLoading = false
    @State private var errorMessage: String?
    @State private var isSubscribed = false

    private let thresholds = [
        (50, "Fair or better", "Any wave is better than no wave"),
        (65, "Good or better", "Recommended — solid sessions"),
        (75, "Great or better", "Worth rescheduling your day"),
        (85, "Epic only", "Rare, but unforgettable")
    ]

    private let subscriptionsKey = "push_subscriptions"

    var body: some View {
        NavigationStack {
            ZStack {
                Color("background").ignoresSafeArea()

                ScrollView {
                    VStack(spacing: 24) {
                        // Header
                        VStack(spacing: 8) {
                            Image(systemName: "bell.badge.waveform.fill")
                                .font(.system(size: 40))
                                .foregroundStyle(.blue)
                            Text("Surf Alerts")
                                .font(.title2.bold())
                                .foregroundStyle(.primary)
                            Text("Get notified when \(spot.name) reaches your target score.")
                                .font(.subheadline)
                                .foregroundStyle(.secondary)
                                .multilineTextAlignment(.center)
                        }
                        .padding(.top, 8)

                        // Permission warning
                        if pushManager.authorizationStatus == .denied {
                            Label("Notifications are disabled. Enable them in Settings.", systemImage: "exclamationmark.triangle.fill")
                                .font(.footnote)
                                .foregroundStyle(.orange)
                                .padding(12)
                                .background(Color.orange.opacity(0.12), in: RoundedRectangle(cornerRadius: 10))
                        }

                        // Threshold picker
                        VStack(alignment: .leading, spacing: 4) {
                            Text("Alert me when the score reaches:")
                                .font(.footnote.weight(.semibold))
                                .foregroundStyle(.secondary)
                                .padding(.horizontal, 4)

                            VStack(spacing: 0) {
                                ForEach(thresholds, id: \.0) { threshold, title, subtitle in
                                    Button {
                                        selectedThreshold = threshold
                                    } label: {
                                        HStack {
                                            VStack(alignment: .leading, spacing: 2) {
                                                Text(title).font(.subheadline.weight(.semibold)).foregroundStyle(.primary)
                                                Text(subtitle).font(.caption).foregroundStyle(.secondary)
                                            }
                                            Spacer()
                                            if selectedThreshold == threshold {
                                                Image(systemName: "checkmark.circle.fill").foregroundStyle(.blue)
                                            } else {
                                                Image(systemName: "circle").foregroundStyle(.tertiary)
                                            }
                                        }
                                        .padding(.horizontal, 16)
                                        .padding(.vertical, 12)
                                        .contentShape(Rectangle())
                                    }
                                    .buttonStyle(.plain)
                                    if threshold != thresholds.last?.0 {
                                        Divider().padding(.leading, 16)
                                    }
                                }
                            }
                            .background(Color("cardBackground"), in: RoundedRectangle(cornerRadius: 14))
                        }

                        // Error
                        if let errorMessage {
                            Text(errorMessage)
                                .font(.footnote)
                                .foregroundStyle(.red)
                                .padding(12)
                                .background(Color.red.opacity(0.1), in: RoundedRectangle(cornerRadius: 10))
                        }

                        // Subscribe / Unsubscribe button
                        Button {
                            Task { await handleToggle() }
                        } label: {
                            HStack {
                                if isLoading {
                                    ProgressView().tint(.white)
                                } else {
                                    Image(systemName: isSubscribed ? "bell.slash.fill" : "bell.fill")
                                    Text(isSubscribed ? "Remove Alert" : "Enable Alert")
                                        .fontWeight(.semibold)
                                }
                            }
                            .frame(maxWidth: .infinity)
                            .padding()
                            .background(isSubscribed ? Color.red.opacity(0.8) : Color.blue)
                            .foregroundStyle(.white)
                            .clipShape(RoundedRectangle(cornerRadius: 14))
                        }
                        .disabled(isLoading || pushManager.authorizationStatus == .denied)

                        Text("You can receive alerts for up to 2 spots. Alerts are sent at most once every 6 hours.")
                            .font(.caption)
                            .foregroundStyle(.tertiary)
                            .multilineTextAlignment(.center)
                    }
                    .padding(20)
                }
            }
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .topBarTrailing) {
                    Button("Done") { isPresented = false }
                }
            }
        }
        .onAppear { checkExistingSubscription() }
    }

    // MARK: - Logic

    private func checkExistingSubscription() {
        let subs = loadSubscriptions()
        isSubscribed = subs.contains { $0.spotId == spot.id }
        if let sub = subs.first(where: { $0.spotId == spot.id }) {
            selectedThreshold = sub.threshold
        }
    }

    private func handleToggle() async {
        if pushManager.deviceToken == nil {
            // Request permission and wait for APNs token
            let granted = await pushManager.requestPermissionAndRegister()
            if !granted {
                errorMessage = "Please enable notifications in Settings to receive surf alerts."
                return
            }
            try? await Task.sleep(for: .seconds(1))
        }

        guard let token = pushManager.deviceToken else {
            errorMessage = "Failed to register for notifications. Try again."
            return
        }

        isLoading = true
        errorMessage = nil
        defer { isLoading = false }

        do {
            if isSubscribed {
                try await APIClient.shared.unsubscribePush(token: token, spotId: spot.id)
                removeSubscription(spotId: spot.id)
                isSubscribed = false
            } else {
                let response = try await APIClient.shared.subscribePush(
                    token: token,
                    spotId: spot.id,
                    threshold: selectedThreshold
                )
                if response.success == true {
                    saveSubscription(spotId: spot.id, spotName: spot.name, threshold: selectedThreshold)
                    isSubscribed = true
                } else {
                    errorMessage = response.error ?? "Failed to subscribe"
                }
            }
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    // MARK: - Persistence

    struct StoredSubscription: Codable {
        let spotId: String
        let spotName: String
        var threshold: Int
    }

    private func loadSubscriptions() -> [StoredSubscription] {
        guard let data = UserDefaults.standard.data(forKey: subscriptionsKey),
              let subs = try? JSONDecoder().decode([StoredSubscription].self, from: data) else { return [] }
        return subs
    }

    private func saveSubscription(spotId: String, spotName: String, threshold: Int) {
        var subs = loadSubscriptions().filter { $0.spotId != spotId }
        subs.append(StoredSubscription(spotId: spotId, spotName: spotName, threshold: threshold))
        if let data = try? JSONEncoder().encode(subs) {
            UserDefaults.standard.set(data, forKey: subscriptionsKey)
        }
    }

    private func removeSubscription(spotId: String) {
        let subs = loadSubscriptions().filter { $0.spotId != spotId }
        if let data = try? JSONEncoder().encode(subs) {
            UserDefaults.standard.set(data, forKey: subscriptionsKey)
        }
    }
}

File: android/app/src/main/java/surf/shouldigo/app/feature/notifications/NotificationSheet.kt
package surf.shouldigo.app.feature.notifications

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import surf.shouldigo.app.data.model.Spot
import surf.shouldigo.app.ui.theme.*

@Composable
fun NotificationSheet(
    spot: Spot,
    onDismiss: () -> Unit,
    viewModel: PushViewModel = hiltViewModel()
) {
    val uiState by viewModel.uiState.collectAsState()

    LaunchedEffect(spot.id) {
        viewModel.loadState(spot.id)
    }

    val thresholds = listOf(
        50 to "Fair or better",
        65 to "Good or better",
        75 to "Great or better",
        85 to "Epic only"
    )

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(24.dp)
    ) {
        // Header
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text(text = "\uD83D\uDD14", fontSize = 24.sp)
            Text(
                text = "Surf Alerts",
                fontSize = 20.sp,
                fontWeight = FontWeight.Bold,
                color = PrimaryText
            )
        }

        Spacer(modifier = Modifier.height(4.dp))
        Text(
            text = "Get notified when ${spot.name} reaches your target score",
            fontSize = 14.sp,
            color = SecondaryText
        )

        if (!uiState.hasPermission) {
            Spacer(modifier = Modifier.height(12.dp))
            Surface(
                shape = RoundedCornerShape(10.dp),
                color = ScoreFair.copy(alpha = 0.1f)
            ) {
                Text(
                    text = "Enable notifications in your device settings to receive alerts",
                    fontSize = 12.sp,
                    color = ScoreFair,
                    modifier = Modifier.padding(12.dp)
                )
            }
        }

        Spacer(modifier = Modifier.height(20.dp))

        // Threshold selection
        Text(
            text = "Alert when score reaches:",
            fontSize = 14.sp,
            fontWeight = FontWeight.SemiBold,
            color = PrimaryText
        )
        Spacer(modifier = Modifier.height(8.dp))

        thresholds.forEach { (value, label) ->
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(vertical = 4.dp),
                verticalAlignment = Alignment.CenterVertically
            ) {
                RadioButton(
                    selected = uiState.threshold == value,
                    onClick = { viewModel.setThreshold(value) },
                    enabled = !uiState.isLoading
                )
                Column(modifier = Modifier.padding(start = 8.dp)) {
                    Text(text = "$value+", fontSize = 15.sp, fontWeight = FontWeight.Medium)
                    Text(text = label, fontSize = 12.sp, color = SecondaryText)
                }
            }
        }

        Spacer(modifier = Modifier.height(20.dp))

        // Subscribe / Unsubscribe button
        if (uiState.isSubscribed) {
            OutlinedButton(
                onClick = { viewModel.unsubscribe(spot.id) },
                modifier = Modifier.fillMaxWidth(),
                shape = RoundedCornerShape(10.dp),
                enabled = !uiState.isLoading
            ) {
                Text("Remove Alert")
            }
        } else {
            val canSubscribe = uiState.subscriptionCount < 2
            Button(
                onClick = { viewModel.subscribe(spot.id, spot.name) },
                modifier = Modifier.fillMaxWidth(),
                shape = RoundedCornerShape(10.dp),
                enabled = !uiState.isLoading && canSubscribe
            ) {
                Text(if (uiState.isLoading) "Setting up..." else "Enable Alert")
            }
            if (!canSubscribe) {
                Spacer(modifier = Modifier.height(4.dp))
                Text(
                    text = "Maximum 2 spot alerts. Remove one first.",
                    fontSize = 12.sp,
                    color = SecondaryText
                )
            }
        }

        uiState.error?.let { error ->
            Spacer(modifier = Modifier.height(8.dp))
            Text(text = error, fontSize = 12.sp, color = ScoreFlat)
        }

        Spacer(modifier = Modifier.height(16.dp))
        Text(
            text = "Alerts are checked periodically and sent at most once every 6 hours per spot.",
            fontSize = 11.sp,
            color = SecondaryText
        )

        Spacer(modifier = Modifier.height(24.dp))
    }
}
